### PR TITLE
Fix RL optimizer offload

### PR DESCRIPTION
**Pull request by jon-barker**
# What does this PR do ?
The current implementation of optimizer cpu offload used in RL does not offload all optimization related state. It uses .cpu() to offload optimizer params and state off the GPU but this doesn't offload grad buffers. There are multiple second order references to the grad buffer tensors, so the memory is not freed by a .cpu() call on only one reference. 

This PR adds new functionality to allow cpu offload of DDP model param and grad buffers and helper functions to use that offload functionality on GPT and MoE DDP models. 

The original optimizer offload functionality is retained as that still correctly handles the non-grad-buffer optimizer state offload.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/rl/rl_utils.py [hunk 1]
```diff
@@ -460,11 +460,6 @@ def get_environment_rollouts(
 
     # If we have seperate training and inference models we to refit weights from the training model to the inference model.
     if inference_model is not None:
-        if args.rl_offload_optimizer_during_inference:
```
Comment (jon-barker) [inline megatron/rl/rl_utils.py:466]: @wdykas can you confirm if it's OK to remove this offload? You added it I think but it actually wasn't doing anything anyway. But I think it's redundant given the other offload call in the inference context

File: megatron/core/distributed/grad_buffer_offload.py [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+    State container holding metadata needed to reallocate grad_data buffers.
+    
+    This is returned by offload_grad_data and should be passed to onload_grad_data.
+    Note: The actual tensor data is NOT preserved - only metadata for reallocation.
+    """
+    # Total number of elements in the buffer (for reallocation)
+    numel: int
+    # Original dtype and device info
+    grad_dtype: torch.dtype
+    original_device: torch.device
```
Comment (yobibyte) [inline megatron/core/distributed/grad_buffer_offload.py]: nit: use consistent naming for both (either original, or drop original)? I like the latter more. Remove comment on 44 too?

File: megatron/core/distributed/grad_buffer_offload.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+allocates fresh buffers. This is useful when the buffer contents are not needed
+(e.g., gradients will be recomputed anyway).
+"""
+
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Tuple
+
+import torch
+
+from .param_and_grad_buffer import _ParamAndGradBuffer, _ParamAndGradBucket, _ParamAndGradBucketGroup
```
Comment (yobibyte) [inline megatron/core/distributed/grad_buffer_offload.py]: Please, clean up unused imports. 

ruff check FNAME --isolated will show those.

File: megatron/core/distributed/grad_buffer_offload.py [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+    bucket_metadata: List[BucketMetadata]
+    # Parameter metadata for restoring param.main_grad views
+    param_grad_metadata: Dict[torch.nn.Parameter, ParamGradMetadata]
+    # Buffer reference (for validation during onload)
+    buffer_id: int
+
+
+def offload_grad_data(
+    ddp_model,
+    optimizer=None,
```
Comment (yobibyte) [inline megatron/core/distributed/grad_buffer_offload.py]: Nit: add typing

File: megatron/core/distributed/grad_buffer_offload.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+    synchronize: bool = True,
+    empty_cache: bool = True,
+) -> List[GradBufferOffloadState]:
+    """
+    Free all grad_data tensors from a DistributedDataParallel model to release GPU memory.
+    
+    This function handles all references to the underlying grad_data tensors:
```
Comment (yobibyte) [inline megatron/core/distributed/grad_buffer_offload.py]: This is an important comment, but belongs to the _offload_single_buffer that actually does this job.

File: megatron/core/distributed/grad_buffer_offload.py [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        bucket.grad_data = buffer.grad_data[start_index:end_index]
+    
+    # Step 3: Recreate param.main_grad views
+    for param, metadata in offload_state.param_grad_metadata.items():
+        param.main_grad = buffer.grad_data[
+            metadata.start_index:metadata.end_index
+        ].view(metadata.shape)
+
+
+def offload_all_grad_data_to_cpu(
```
Comment (yobibyte) [inline megatron/core/distributed/grad_buffer_offload.py]: What is a purpose of this function?

File: megatron/core/distributed/grad_buffer_offload.py [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+    """
+    Convenience function to free all grad buffers from GPU.
+    
+    Alias for offload_grad_data with default parameters.
+    Note: Despite the name, this does NOT copy data to CPU - it only frees GPU memory.
+    """
+    return offload_grad_data(ddp_model, optimizer)
+
+
+def onload_all_grad_data_from_cpu(
```
Comment (yobibyte) [inline megatron/core/distributed/grad_buffer_offload.py]: What is a purpose of this function? Why cannot we just call onload_grad_data directly?

File: megatron/core/distributed/grad_buffer_offload.py [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+                shape=param.main_grad.shape,
+            )
+    
+    # Step 2: Clear cached shard lists in bucket groups that reference this buffer's buckets
+    buffer_bucket_set = set(buffer.buckets)
+    for bucket_group in bucket_groups:
+        # Check if this bucket group contains buckets from this buffer
+        if any(b in buffer_bucket_set for b in bucket_group.buckets):
+            for i in range(len(bucket_group.cached_grad_buffer_shard_list)):
+                bucket_group.cached_grad_buffer_shard_list[i] = None
```
Comment (yobibyte) [inline megatron/core/distributed/grad_buffer_offload.py]: I am not totally sure what is going on here, but it looks sus that we check any(), but clear **every** cached_grad_buffer_shard_list.

File: megatron/core/distributed/grad_buffer_offload.py [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+def onload_grad_data(
+    ddp_model,
+    offload_states: List[GradBufferOffloadState],
+    optimizer=None,
+    synchronize: bool = True,
+) -> None:
+    """
+    Reallocate grad_data tensors on GPU.
+    
+    This function allocates fresh buffers and recreates all references:
```
Comment (yobibyte) [inline megatron/core/distributed/grad_buffer_offload.py]: This also belongs to the _onload_single_buffer().

File: megatron/core/distributed/grad_buffer_offload.py [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+    1. The main buffer.grad_data tensor (freshly allocated, uninitialized)
+    2. bucket.grad_data views in each bucket
+    3. param.main_grad views on each parameter
```
Comment (yobibyte) [inline megatron/core/distributed/grad_buffer_offload.py]: We offloaded 4 things, but onload 3, is this by design?

File: megatron/core/distributed/grad_buffer_offload.py [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+    
+    # Step 1: Allocate fresh tensor on GPU (contents are uninitialized)
+    buffer.grad_data = torch.empty(
+        offload_state.numel,
+        dtype=offload_state.grad_dtype,
+        device=offload_state.original_device,
+    )
+    
+    # Step 2: Recreate bucket grad_data views
+    for bucket, metadata in zip(buffer.buckets, offload_state.bucket_metadata):
```
Comment (yobibyte) [inline megatron/core/distributed/grad_buffer_offload.py]: add strict=True?

File: megatron/core/distributed/grad_buffer_offload.py [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+class BucketMetadata:
+    """Metadata needed to restore a bucket's grad_data view."""
+    offset: int
+    numel: int
+
+
+@dataclass
+class ParamGradMetadata:
+    """Metadata needed to restore a parameter's main_grad view."""
+    start_index: int
```
Comment (yobibyte) [inline megatron/core/distributed/grad_buffer_offload.py]: Not entirely sure here, but it looks like the logic of this metadata is equivalent to BucketMetadata, should we also name those `offset` and `numel` or the other way round?

File: megatron/core/distributed/grad_buffer_offload.py [hunk 13]
```diff
@@ -0,0 +1,309 @@
+# Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
```
Comment (yobibyte) [inline megatron/core/distributed/grad_buffer_offload.py]: 2026?

File: megatron/rl/rl_utils.py (hunk 1 above)
Comment (wdykas) [reply megatron/rl/rl_utils.py:466]: I dont think this is redundant. The problem is we moved optimizer offload into the inference context manager megatron_rl_inference_mode.  When we have refit we have a separate inference and training model so the inference model get called with megatron_rl_inference_mode but that mode isnt the one with the optimizer state. So we need this call when doing refit to actually offload the optimizer

Comment (wdykas) [reply megatron/rl/rl_utils.py:466]: I admit its a bit strange of a pattern but I think this is needed unless we want to fully move the optimizer offload out of that context.

File: megatron/core/distributed/grad_buffer_offload.py (hunk 3 above)
Comment (jon-barker) [reply megatron/core/distributed/grad_buffer_offload.py]: done

File: megatron/rl/rl_utils.py (hunk 1 above)
Comment (jon-barker) [reply megatron/rl/rl_utils.py:466]: OK - makes sense. Then I need to replace this with the new functions for optimizer offload

File: megatron/core/distributed/grad_buffer_offload.py (hunk 12 above)
Comment (jon-barker) [reply megatron/core/distributed/grad_buffer_offload.py]: The naming is intentional and consistent.

`BucketMetadata.offset/numel` matches how `_ParamAndGradBucket` parameterizes its position.

`ParamGradMetadata.start_index/end_index` matches how `param_index_map` stores parameter locations.

I think this is more important than consistency within just this file.

File: megatron/core/distributed/grad_buffer_offload.py (hunk 2 above)
Comment (jon-barker) [reply megatron/core/distributed/grad_buffer_offload.py]: done

File: megatron/core/distributed/grad_buffer_offload.py (hunk 4 above)
Comment (jon-barker) [reply megatron/core/distributed/grad_buffer_offload.py]: done - this actually helped me spot a gap in the implementation, we weren't iterating over ChainedOptimizers correctly

File: megatron/core/distributed/grad_buffer_offload.py (hunk 10 above)
Comment (jon-barker) [reply megatron/core/distributed/grad_buffer_offload.py]: yes - it says in the comment that "cached_grad_buffer_shard_list will be recreated lazily on next use"

File: megatron/core/distributed/grad_buffer_offload.py (hunk 11 above)
Comment (jon-barker) [reply megatron/core/distributed/grad_buffer_offload.py]: good idea - added

File: megatron/core/distributed/grad_buffer_offload.py (hunk 13 above)
Comment (jon-barker) [reply megatron/core/distributed/grad_buffer_offload.py]: yep, done

File: megatron/core/distributed/grad_buffer_offload.py (hunk 6 above)
Comment (jon-barker) [reply megatron/core/distributed/grad_buffer_offload.py]: it's legacy from an earlier version - deleting it.

**Comment by jon-barker** [comment]
> Thanks for adding this, @jon-barker. I left some comments, mostly nits.
> 
> What do you think about adding the following test:
> 
> * We create a model, with @tdene's changes, we can make a real GPT model as in my prepare_data unit test.
> * We measure the used memory.
> * We offload.
> * We measure the used/offloaded memory.
> * We onload.
> * We measure the used/offloaded memory.
> 
> The deltas should be roughly the same, i.e. the size of the offloaded should be about post onload used - post offload used?

Yes, was planning to implement a test of this form this morning.

File: megatron/core/distributed/grad_buffer_offload.py (hunk 12 above)
Comment (yobibyte) [reply megatron/core/distributed/grad_buffer_offload.py]: agreed

**Comment by jon-barker** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3112/commits/94148d5a390e9fabcbcbb5af841b21c7f48a073f

**Comment by jon-barker** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3112/commits/598c6b5818d7cae10ab40c2b8b4baf4540a6550b

File: megatron/core/distributed/grad_buffer_offload.py (hunk 5 above)
Comment (jon-barker) [reply megatron/core/distributed/grad_buffer_offload.py]: split the comment across the functions

File: megatron/core/distributed/grad_buffer_offload.py (hunk 9 above)
Comment (jon-barker) [reply megatron/core/distributed/grad_buffer_offload.py]: split the comment across the functions

File: megatron/core/distributed/grad_buffer_offload.py (hunk 7 above)
Comment (jon-barker) [reply megatron/core/distributed/grad_buffer_offload.py]: deleted this function, it's unecesary, it was legacy from an earlier implementation

File: megatron/core/distributed/grad_buffer_offload.py (hunk 8 above)
Comment (jon-barker) [reply megatron/core/distributed/grad_buffer_offload.py]: Good catch - i think this could be a bug for example with FP8 where we'd have buckets from multiple buffers

**Comment by jon-barker** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3112/commits/1638f3e10fb377412d49459fd40541fcc49222d8

**Comment by jon-barker** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3112/commits/95ce3f9b5cf92b164934c80d19d3051c82dc0c65

File: megatron/core/distributed/grad_buffer_offload.py (hunk 8 above)
Comment (yobibyte) [reply megatron/core/distributed/grad_buffer_offload.py]: perfect! closing this

**Comment by jon-barker** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3112/commits/1d650b5bd14a7260f52c72ed12fa450570d540fe

**Comment by jon-barker** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3112/commits/caf4aaeefda18415b9deecb813f5b45172dd93de

**Comment by jon-barker** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3112/commits/73aa953598a6a30220dc45804e14d359e1016aaa

File: megatron/rl/rl_utils.py (hunk 1 above)
Comment (yobibyte) [reply megatron/rl/rl_utils.py:466]: I think it makes a lot of sense to turn this conversation into a comment and keep it in the code for the future generation not to ask this Q again/not to remove by accident.

**Comment by jon-barker** [comment]
> Nice! Thanks for addressing my comments and adding tests. LGTM, but let's add your conversation with @wdykas as a comment.

Makes sense - here's a link to the relevant conversation: https://github.com/NVIDIA/Megatron-LM/pull/3112#discussion_r2734360520

I have also updated the PR description to better describe what's happening here.

**Review by jaredcasper — APPROVED** [review]

**Comment by jon-barker** [comment]
Turns out there is a simpler approach for freeing the grad buffer memory proposed in https://github.com/NVIDIA/Megatron-LM/pull/2951 - I am going to refactor this PR to adopt their approach.

**Comment by jon-barker** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3112/commits/3e56e0f1a447d9870cc680904d8a217a5da8951c

**Comment by jon-barker** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3112/commits/4030c8c23fb435f9e611fb885c467080c30518a5

File: tests/unit_tests/rl/test_rl_utils.py [hunk 14]
```diff
@@ -416,3 +424,156 @@ def test_prepare_trajectories(self, use_sequence_packing):
                 got_t = got if torch.is_tensor(got) else torch.tensor(got, dtype=torch.float32)
                 exp_t = torch.tensor(exp, dtype=torch.float32, device=got_t.device)
                 torch.testing.assert_close(got_t, exp_t, rtol=0, atol=0)
+
+    def test_grad_buffer_offload(self):
```
Comment (tdene) [inline tests/unit_tests/rl/test_rl_utils.py]: Could you please avoid using `Utils.initialize_model_parallel()` and `Utils.destroy_model_parallel()` directly, but instead use the `initialize_model_parallel` fixture?

Comment (jon-barker) [reply tests/unit_tests/rl/test_rl_utils.py]: done

**Comment by jon-barker** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3112/commits/170f3035bc6c6ac12f675f0500964ba1d7d9e4ff

**Comment by jon-barker** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3112/commits/f71cc38a9c2d5619d4f13acf51e4eff603db465a

**Comment by jon-barker** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3112/commits/3749d109c268f3b202bd62fd29cd0be6a5d0987f

**Review by guyueh1 — APPROVED** [review]
LGTM
